x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpoint.java

```diff
@@ -0,0 +1,259 @@
+    private static final ConstructingObjectParser<DataFrameTransformCheckpoint, Void> LENIENT_PARSER = createParser(true);
+
+    private final String id;
+    private final long timestampMillis;
+    private final long checkpoint;
+    private final Map<String, long[]> indicesCheckpoints;
+    private final long timeUpperBoundMillis;
+
+    private static ConstructingObjectParser<DataFrameTransformCheckpoint, Void> createParser(boolean lenient) {
+        ConstructingObjectParser<DataFrameTransformCheckpoint, Void> parser = new ConstructingObjectParser<>(NAME,
+                lenient, args -> {
+                    String id = (String) args[0];
+                    Long timestamp = (Long) args[1];
+                    Long checkpoint = (Long) args[2];
+
+                    @SuppressWarnings("unchecked")
+                    Map<String, long[]> checkpoints = (Map<String, long[]>) args[3];
+
+                    Long timestamp_checkpoint = (Long) args[4];
+
+                    // ignored, only for internal storage: String docType = (String) args[5];
+                    return new DataFrameTransformCheckpoint(id, timestamp, checkpoint, checkpoints, timestamp_checkpoint);
+                });
+
+        parser.declareString(constructorArg(), DataFrameField.ID);
+
+        // note: this is never parsed from the outside where timestamp can be formatted as date time
+        parser.declareLong(constructorArg(), TIMESTAMP_MILLIS);
+        parser.declareLong(constructorArg(), CHECKPOINT);
+
+        parser.declareObject(constructorArg(), (p,c) -> {
+            Map<String, long[]> checkPointsByIndexName = new TreeMap<>();
+            XContentParser.Token token = null;
+            while ((token = p.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token != XContentParser.Token.FIELD_NAME) {
+                    throw new ParsingException(p.getTokenLocation(), "Unexpected token " + token + " ");
+                }
+
+                final String indexName = p.currentName();
+                token = p.nextToken();
+                if (token != XContentParser.Token.START_ARRAY) {
+                    throw new ParsingException(p.getTokenLocation(), "Unexpected token " + token + " ");
+                }
+
+                long[] checkpoints = p.listOrderedMap().stream().mapToLong(num -> ((Number) num).longValue()).toArray();
+                checkPointsByIndexName.put(indexName, checkpoints);
+            }
+            return checkPointsByIndexName;
+        }, INDICES);
+        parser.declareLong(optionalConstructorArg(), TIME_UPPER_BOUND_MILLIS);
+        parser.declareString(optionalConstructorArg(), DataFrameField.INDEX_DOC_TYPE);
+
+        return parser;
+    }
+
+    public DataFrameTransformCheckpoint(String id, Long timestamp, Long checkpoint, Map<String, long[]> checkpoints, Long timeUpperBound) {
+        this.id = id;
```

Comment:
`id` is the Data Frame Transform id? So the delete by query in delete transform will clean up these checkpoint docs?

Reply:
yes

Reply:
Can you rename to `transformId` or similar so that its clear what it refers to please

Reply:
I agree that changing it to `transformId` is preferable.